README.md
# PS2_Mouse_to_Serial_Port
Un convertidor para conectar un ratón PS/2 a un puerto serie (RS-232).


1. Introducción
--------------------------------------------------------------------------------
Los ratones con conexión DB9 para puerto serie cada vez son más difíciles de encontrar (y más caros), lo que puede suponer un problema si necesitamos emplear uno en un ordenador clásico que solo disponga del puerto serie para su conexión. Después de investigar unas horas, encontré varias opciones comerciales y un par de proyectos open source que, usando placas Arduino, permitían la conexión de un ratón PS/2, o USB compatible con el protocolo PS/2 a un puerto serie.

https://github.com/Lameguy64/ps2serial
 
https://github.com/kristopher/PS2-Mouse-Arduino

Una vez documentado sobre estos proyectos y realizando un primer prototipo funcional de los mismos, decidí realizar mi propia versión del conversor.


2. Como funciona
--------------------------------------------------------------------------------
Básicamente la placa Arduino se encarga de prácticamente todo el trabajo, necesitando solo un poco de electrónica externa para su funcionamiento. Una librería de interacción con el protocolo PS/2 no permite leer la señal codificada procedente del ratón con las entradas. Dicha información se codifica en un paquete de datos en el formato que esperan los drivers de ratón mas comunes y se envían al puerto serie. Además, se monitorea la señal RTS del puerto serie por si se solicita el reinicio del envío de datos.


3. Hardware usado en el prototipo
--------------------------------------------------------------------------------
Se han realizado dos prototipos de prueba, usando las placas Arduino UNO R3 y Arduino NANO, funcionando ambas placas perfectamente. El resto de electrónica sirve para la adaptación de las señales que circulan por el puerto serie, ofreciendo cierto nivel de aislamiento y protección entre este puerto y la placa Arduino. Además, se incorporan dos diodos LED indicadores de estado, para el monitoreo del comportamiento del prototipo.


4. Que indican los LED’s de estado
--------------------------------------------------------------------------------
En mi caso he usado un LED rojo para el indicador de alimentación y un LED amarillo para el indicador de estado. Dicho indicador nos proporciona cierta información en el momento del arranque del adaptador, así como de su funcionamiento. Al encender el adaptador, si no puede detectar un ratón compatible conectado, quedará encendido de manera permanente. Dicho LED de estado parpadeará de manera intermitente una vez por segundo si ha detectado un ratón, pero no es posible iniciar la comunicación siguiendo el protocolo PS/2. Si el proceso de inicio se realiza correctamente, este LED emitirá 3 parpadeos, uno largo seguido de 2 mas cortos. Una vez iniciado el adaptador, si movemos el ratón, observaremos que este LED parpadea regularmente, quedando encendido mientras mantengamos pulsado alguno de los 3 botones del ratón. Si movemos la rueda de scroll, tambien emitira un parpadeo.
Si se recibe desde el puerto serie la señal RTS, este LED emitirá un parpadeo de aproximadamente medio segundo. Esto nos permite verificar que el driver ha detectado correctamente nuestro adaptador.


5. Capacidades extendidas
--------------------------------------------------------------------------------
Dado que este proyecto (en especial la placa ISA) está destinado a ordenadores sin puerto PS/2 (normalmente 486 y anteriores), usando el sistema operativo MS-DOS o Windows 3.11 y anteriores, junto con el hecho de que actualmente casi todos los ratones PS/2 son ópticos y disponen de la rueda de scroll, sumándole que para estos ordenadores casi ningún programa hace uso de la rueda de scroll, he decidido usar dicha rueda para ajustar a tiempo real la velocidad de desplazamiento del cursor. Dado que, según la resolución de pantalla, programa o modelo de ratón esta puede ser diferente, creo que es una característica útil. La velocidad seleccionada se conservará hasta que apaguemos el adaptador (o el ordenador en el caso de la placa ISA), no viendose afectada por el reinicio del sistema. Si queremos guardar la velocidad actual de manera permanente, deberemos de mantener pulsados los botones izquierdo y derecho de manera simultanea durante 5 segundos, observando que el LED de estado se apaga transcurrido ese tiempo. Esto nos indica que la configuración de ha almacenado de manera correcta. Si en algún momento deseamos restaurar la configuración por defecto, deberemos de mantener presionados ambos botones durante 10 segundos, el LED de estado se apagará a los 5 segundos, volviéndose a encender transcurridos 10 segundos. Esto indicará que la configuración se ha restaurado y almacenado con éxito.

6. Precauciones
--------------------------------------------------------------------------------
- Conectar y desconectar siempre el ratón al puerto PS/2 con el adaptador desconectado. La conexión o desconexión del ratón con el adaptador conectado a la alimentación podría causar daños al ratón y al propio adaptador. Esto también se aplica a la conexión del adaptador con el puerto serie del ordenador, siendo muy recomendable que todas las conexiones se realicen con los equipos apagados.
- Cualquier dispositivo conectado al puerto RS-232 del ordenador puede causar daños a dicho puerto o al dispositivo si este tiene algún error de montaje, por lo que se recomienda que repaséis 3 veces las conexiones antes de conectarlo a vuestro ordenador.
- Aplicable a la versión ISA, en ningún caso actualicéis el firmware del modulo Arduino con la tarjeta ISA insertada en la placa base. Retirar la tarjeta ISA o el módulo Arduino antes de conectarlo al puerto USB para su actualización. De no hacerlo, se corre el riesgo de dañar el módulo Arduino la propia placa base del ordenador.


docs/changelog.txt
/*******************************************************************************
  
    Adaptador de mouse PS2 a puerto serie RS-232
	
	--- Archivo de registro de cambios ---
	

    Proyecto iniciado el Martes, 25 de Julio del 2023
    por Cesar Rincon "NightFox"

    https://nightfoxandco.com
    contact@nightfoxandco.com

    "Adaptador de mouse PS2 a puerto serie RS-232" is under MIT License
    Copyright (c) 2023 by Cesar Rincon "NightFox"

*******************************************************************************/

2023-09-06 V1.8.0
--------------------------------------------------------------------------------
- Dado que algunos modelos de ratón no funcionaban como es debido y en algunos
  casos ni encendían (PERIMICE-209), se analiza la librería usada de Arduino
  para la interfaz con el puerto PS/2. Viendo que no seguía el protocolo
  estándar PS/2, se escribe una librería con el mínimo necesario para
  interactuar con un ratón PS/2, siguiendo las normas de dicho protocolo.
- Ahora todos los comandos que se envían al ratón se verifican.
- Se añade una secuencia de encendido para activar algunos modelos de ratón
  que antes no funcionaban, como el PERIMICE-209.
- Con la librería personalizada, se añade la opción de detectar si el ratón es
  compatible con Microsoft IntelliMouse y este dispone de características
  extendidas, como la rueda de scroll. De ser así, se activa la función a
  tiempo real de cambio de velocidad del cursor usando la rueda de scroll.
- Se añade la función de guardado de la velocidad actual del cursor
  (manteniendo pulsados los botones derecho e izquierdo durante 5 segundos) y
  la función de restauración de la velocidad por defecto (manteniendo pulsados
  los botones derecho e izquierdo durante 5 segundos).
- Siguiendo el protocolo estándar PS/2, a partir de ahora no es posible la
  conexión en caliente del ratón al puerto PS/2.
- Se añaden nuevas indicaciones de estado en el LED correspondiente.


2023-08-29 V1.7.1
--------------------------------------------------------------------------------
- Modificado el ancho de la pista de alimentación del puerto PS/2
  de la placa ISA.


2023-08-28 V1.7.0
--------------------------------------------------------------------------------
- Modificadas las huellas de todas las placas para que los componentes TO92
  resulten más fáciles de soldar a mano.
- Modificadas las conexiones del conector IDC en la placa ISA para poder usar
  un cable IDC a DB9 comercial, sin necesidad de realizarle modificaciones.
- Añadido el modelo en formato FreeCad y el archivo STL del soporte para la
  placa ISA. Estos archivos estan preparados para poder realizar la impresión
  en 3D del soporte. Se recomienda un in-fill del 100%.
  

2023-08-24 V1.6.0
--------------------------------------------------------------------------------
- Modificada la librería PS2Mouse para evitar que la placa Arduino se quede
  colgada al iniciarse si no se detecta el mouse conectado en el puerto PS/2.
  Ahora si la inicialización del mouse falla, se reescanea el puerto a cada
  segundo. De todos modos, os recomiendo que no conectéis o desconectéis el
  mouse con el adaptador/ordenador encendido, ya que hay la posibilidad de que
  se causen daños al adaptador o al propio mouse.


2023-08-14 V1.5.1
--------------------------------------------------------------------------------
- Arreglada la serigrafía del conector MINI-DIN-6 (PS/2), la cual se
  superponía en algunas zonas de soldado.
- En todas las placas se añade el símbolo de los agujeros de montaje con el
  tamaño estándar M3.


2023-08-13 V1.5.0
--------------------------------------------------------------------------------
- Ajustadas las placas "Stand alone" y "In-a-box" al tamaño standard de
  90x70 mm.
- Añadidas a estas placas las zonas de relleno a masa (GND).
- Añadidas las serigrafias con los valores de todos los componentes en todas
  las versiones de las placas.


2023-08-12 V1.4.0
--------------------------------------------------------------------------------
- Pequeños ajustes al tamaño y contorno de la placa ISA.
- Se han añadido zonas de relleno para GND en la placa ISA.
- Sin cambios en el software.
- Importante! Recordad que aún no he podido verificar en "físico" el diseño
  de las placas, por lo que recomiendo precaución si van a ser usadas.


2023-08-12 V1.3.0
--------------------------------------------------------------------------------
- Se utiliza una referencia de versión para el proyecto, independientemente
  de las versiones de placa o software para facilitar la identificación de
  la versión actual.
- Se añade el esquema y diseño de placa para la versión "ISA".
- Se amplía el ancho de pista en las placas restantes.
- Sin cambios en el software.
- Importante! Recordad que aún no he podido verificar en "físico" el diseño
  de las placas, por lo que recomiendo precaución si van a ser usadas.


2023-08-11 V1.0.1-rev2.1
--------------------------------------------------------------------------------
- Corregido un error en la colocación del conector PS/2 en la placa
  "Stand Alone" del proyecto KiCad.
- Recordar que aún no he podido producir físicamente estas placas, por lo que
  si vais a producirlas, revisar el proyecto 3 veces en búsqueda de errores.

2023-08-10 V1.0.1-rev2
--------------------------------------------------------------------------------
- Se añade el esquema realizado en formato KiCad.
- Corregido un error en la conexión a masa del condensador C6 en el esquema
  en formato TinyCad.
- Modificadas algunas de las anotaciones en los pins de Arduino en el esquema
  en formato TinyCad.
- Añadidos los esquemas de KiCad en formato PDF y PNG.
- Añadidos los diseños de las PBC para las version "Stand Alone" y "In-a-Box"
  Estos diseños se han mandado a producción, pero aún no se han verificado.
- No hay cambios en el código.


2023-08-08 V1.0.0-stable
--------------------------------------------------------------------------------

arduino_project/ps2_to_serial/src/config.h
/*******************************************************************************
  
    Adaptador de mouse PS2 a puerto serie RS-232

    Datos de configuracion

    Proyecto iniciado el Martes, 25 de Julio del 2023
    por Cesar Rincon "NightFox"

    https://nightfoxandco.com
    contact@nightfoxandco.com

    Basado en los siguientes proyectos de GitHub
    https://github.com/Lameguy64/ps2serial
    https://github.com/kristopher/PS2-Mouse-Arduino
  

    "Adaptador de mouse PS2 a puerto serie RS-232" is under MIT License

    Copyright (c) 2023 by Cesar Rincon "NightFox"

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal	in the Software without restriction,
    including without limitation the rights to use, copy, modify, merge,
    publish, distribute, sublicense, and/or sell copies of the Software,
    and to permit persons to whom the Software is furnished to do so,
    subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all	copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER	LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
*******************************************************************************/



#ifndef __CONFIG_H
#define __CONFIG_H


/*** Includes ***/
#include <Arduino.h>



/*** Define los PINS a usar en la placa Arduino ***/
const uint8_t PIN_MOUSE_CLOCK = 2;          // Señal de reloj del protocolo PS/2 (Pin 5 del conector PS/2)
const uint8_t PIN_MOUSE_DATA = 5;           // Datos del protocolo PS/2 (Pin 1 del conector PS/2)

const uint8_t PIN_RTS_SIGNAL = 7;           // Señal RTS del puerto serie (Pin 7 del conector DB9) *1

const uint8_t PIN_MSG_LED = 8;              // LED indicador de estado

// *1 - No conectar NUNCA este pin del puerto serie directamente a un pin de la placa Arduino



/*** Define los datos de configuracion del protocolo PS/2 ***/
const uint8_t READ_LOOPS = 4;               // Numero de lecturas del puerto PS/2 por ciclo [4]
const uint8_t SAMPLE_RATE = 100;            // Samples/seg [10, 20, 40, 60, 80, >> 100 <<, 200]
const uint8_t RESOLUTION = 0x02;            // Resolucion [0x00 = 1 count/mm, 0x01 = 2 count/mm, >> 0x02 = 4 count/mm <<, 0x03 = 8 count/mm]


/*** Define los datos de configuracion del protocolo SERIE ***/
const uint32_t SERIAL_BAUD_RATE = 1200;
const uint8_t SERIAL_OPTIONS = SERIAL_7N1;


/*** Configuracion adicional del mouse ***/
const int8_t MOUSE_BLINK_DELAY = 3;         // Espera entre parpadeos [3]
const float MOUSE_DEFAULT_SPEED = 0.5f;     // Factor de escala en la velocidad [0.5f]
const float MOUSE_SPEED_CHANGE = 0.05f;     // Factor de cambio [0.05f]
const float MOUSE_MIN_SPEED = 0.1f;         // Velocidad minima [0.1f]
const float MOUSE_MAX_SPEED = 1.0f;         // Valocidad maxima [1.0f]
const float SAVE_HELD_TIME = 5000;          // Tiempo que se requiere para el guardado de datos en la eeprom [5s]
const float RESTORE_HELD_TIME = 10000;       // Tiempo que se requiere para la restauracion y guardado de datos en la eeprom [10s]
const uint8_t CHK0 = 0x01, CHK2 = 0x04;     // Checksum de datos de la eeprom 


#endif


arduino_project/ps2_to_serial/src/maincode.h
/*******************************************************************************
  
    Adaptador de mouse PS2 a puerto serie RS-232

    Clase principal del proyecto

    Proyecto iniciado el Martes, 25 de Julio del 2023
    por Cesar Rincon "NightFox"

    https://nightfoxandco.com
    contact@nightfoxandco.com

    Basado en los siguientes proyectos de GitHub
    https://github.com/Lameguy64/ps2serial
    https://github.com/kristopher/PS2-Mouse-Arduino
  

    "Adaptador de mouse PS2 a puerto serie RS-232" is under MIT License

    Copyright (c) 2023 by Cesar Rincon "NightFox"

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal	in the Software without restriction,
    including without limitation the rights to use, copy, modify, merge,
    publish, distribute, sublicense, and/or sell copies of the Software,
    and to permit persons to whom the Software is furnished to do so,
    subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all	copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER	LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
*******************************************************************************/



#ifndef __MAINCODE_H
#define __MAINCODE_H


/*** Includes ***/
// Arduino
#include <Arduino.h>
// Proyecto
#include "config.h"
#include "ps2_mouse_interface.h"



/*** Declaracion de la clase ***/
class MainCode {


    public:

        // Contructor & destructor
        MainCode();
        ~MainCode();

        // Metodo para la configuracion inicial
        void Start();

        // Metodo del bucle de ejecucion del programa
        void Update();


    private:

        // Objeto de la clase PS2Mouse (Acceso al raton PS/2)
        Ps2MouseInterface ps2mouse = Ps2MouseInterface(PIN_MOUSE_CLOCK, PIN_MOUSE_DATA, SAMPLE_RATE, RESOLUTION);

        // Variables de control de los datos del raton PS/2
        struct {
            struct {
                bool left;
                bool right;
                bool middle;
            } button;
            struct {
                bool left;
                bool right;
                bool middle;
                int32_t wheel;
            } last;
            struct {
                int32_t x;
                int32_t y;
                int32_t wheel;
            } position;
            struct {
                int32_t x;
                int32_t y;
                int32_t wheel;
            } delta;
            uint8_t packet[4];
            uint8_t packet_size;
            int16_t data[4];            // [0] - Status bits, [1] - Movement X, [2] - Movement Y, [4] - Movement Z
            bool event;                 // Evento de raton
            int8_t led;                 // Control del led de actividad
            float speed;                // Velocidad del cursor
            int32_t save_timer;         // Temporizador para el guardado
            bool save_flag;             // Flag de guardado
            int32_t restore_timer;      // Temporizador para la restauracion
            bool restore_flag;          // Flag para la restauracion
        } mouse;

        // Variables de control del puerto RS-232
        struct {
            bool rts;
        } rs232;

        // Metodo para establecer los valores por defecto
        void SetDefaults();
        // Metodo para configurar los PINS de la placa Arduino
        void SetPins();
        // Metodo para configurar el protocolo PS/2
        void SetPs2();
        // Metodo para configurar el protocolo SERIE de la placa Arduino
        void SetSerial();

        // Lectura de los datos del Raton PS/2
        void ReadMouse();
        // Analiza y prepara los datos para su envio
        void ParseData();
        // Led de actividad
        void ActivityLed();
        // Codifica el paquete de datos
        void EncodeData();
        // Envia los datos al puerto SERIE
        void SendData();
        // Monitor de la señal RTS del puerto serie
        void RtsStatusMonitor();

        // Guarda la configuracion de velocidad
        void SaveMouseSpeed();
        // Lee la configuracion de velocidad
        void LoadMouseSpeed();

};



#endif


arduino_project/ps2_to_serial/src/maincode.cpp
/*******************************************************************************
  
    Adaptador de mouse PS2 a puerto serie RS-232

    Clase principal del proyecto

    Proyecto iniciado el Martes, 25 de Julio del 2023
    por Cesar Rincon "NightFox"

    https://nightfoxandco.com
    contact@nightfoxandco.com

    Basado en los siguientes proyectos de GitHub
    https://github.com/Lameguy64/ps2serial
    https://github.com/kristopher/PS2-Mouse-Arduino
  

    "Adaptador de mouse PS2 a puerto serie RS-232" is under MIT License

    Copyright (c) 2023 by Cesar Rincon "NightFox"

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the "Software"), to deal	in the Software without restriction,
    including without limitation the rights to use, copy, modify, merge,
    publish, distribute, sublicense, and/or sell copies of the Software,
    and to permit persons to whom the Software is furnished to do so,
    subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all	copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER	LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
*******************************************************************************/



/*** Includes ***/
// Arduino
#include <Arduino.h>
#include <EEPROM.h>
// Proyecto
#include "config.h"
#include "ps2_mouse_interface.h"
#include "maincode.h"



/*** Constructor ***/
MainCode::MainCode() {

}



/*** Destructor ***/
MainCode::~MainCode() {

}



/*** Metodo para la configuracion inicial ***/
void MainCode::Start() {

    // Establece los valores por defecto
    SetDefaults();
    // Establece los pins
    SetPins();
    // Indica al raton PS/2 que debe encenderse
    ps2mouse.PowerUp();
    // Inicia el protocolo PS/2
    SetPs2();
    // Inicia el protocolo SERIE
    SetSerial();
    // Carga la configuracion por defecto del raton (velocidad)
    LoadMouseSpeed();

    // 3 parpadeos para indicar el bootup correcto
    for (uint8_t i = 0; i < 3; i ++) {
        digitalWrite(PIN_MSG_LED, HIGH);
        delay(100);
        digitalWrite(PIN_MSG_LED, LOW);
        delay(100);
    }
 
}



/*** Metodo del bucle de ejecucion del programa ***/
void MainCode::Update() {

    // Prepara los datos para esta iteracion
    mouse.button.left = false;
    mouse.button.right = false;
    mouse.button.middle = false;
    mouse.event = false;

    // Lectura de los datos del puerto PS/2
    // Dado que el puerto PS/2 es mucho mas rapido que el puerto SERIE,
    // estos datos se leeran "READ_LOOPS" veces para compensar la diferencia de velocidades
    for (uint8_t i = 0; i < READ_LOOPS; i ++) ReadMouse();

    // Analisis y preparacion de los datos recogidos
    ParseData();
    // LED de actividad
    ActivityLed();
    // Codifica los datos recogidos
    EncodeData();
    // Envia, si es necesario, el paquete de datos por el puero serie
    SendData();
    // Monitoreo de la señal RTS
    RtsStatusMonitor();

    // Guardado de datos?
    SaveMouseSpeed();

    // Registra el estado de los botones para la siguiente iteracion
    mouse.last.left = mouse.button.left;
    mouse.last.right = mouse.button.right;
    mouse.last.middle = mouse.button.middle;
    mouse.last.wheel = mouse.delta.wheel;

}



/*** Metodo para establecer los valores por defecto ***/
void MainCode::SetDefaults() {

    // Mouse
    mouse.position.x = 0;
    mouse.position.y = 0;
    mouse.position.wheel = 0;
    mouse.delta.x = 0;
    mouse.delta.y = 0;
    mouse.delta.wheel = 0;
    mouse.button.left = false;
    mouse.button.right = false;
    mouse.button.middle = false;
    mouse.last.left = false;
    mouse.last.right = false;
    mouse.last.middle = false;
    mouse.last.wheel = 0;
    for (uint8_t i = 0; i < 4; i ++) mouse.packet[i] = 0;
    mouse.packet_size = 3;
    mouse.event = false;
    mouse.led = 0;
    mouse.speed = MOUSE_DEFAULT_SPEED;
    mouse.save_timer = (millis() + SAVE_HELD_TIME);
    mouse.save_flag = false;
    mouse.restore_timer = (millis() + RESTORE_HELD_TIME);
    mouse.restore_flag = false;

    // RS-232
    rs232.rts = false;

}



/*** Metodo para configurar los PINS de la placa Arduino ***/
void MainCode::SetPins() {

    // Señal RTS del puerto serie
    pinMode(PIN_RTS_SIGNAL, INPUT);

    // LED de estado
    pinMode(PIN_MSG_LED, OUTPUT);
    digitalWrite(PIN_MSG_LED, LOW);

}



/*** Metodo para configurar el protocolo PS/2 ***/
void MainCode::SetPs2() {

    digitalWrite(PIN_MSG_LED, HIGH);
    
    bool ps2_ready = false;
    do {
        // Intenta iniciar el mouse ps/2, si no se consigue, reintentalo en 1 segundo
        ps2_ready |= ps2mouse.Init();
        if (!ps2_ready) {
            delay(500);
            digitalWrite(PIN_MSG_LED, LOW);
            delay(500);
            digitalWrite(PIN_MSG_LED, HIGH);
        }
    } while (!ps2_ready);

    digitalWrite(PIN_MSG_LED, LOW);

}



/*** Metodo para configurar el protocolo SERIE de la placa Arduino ***/
void MainCode::SetSerial() {

    digitalWrite(PIN_MSG_LED, HIGH);

    Serial.begin(SERIAL_BAUD_RATE, SERIAL_OPTIONS);
    while(!Serial);

    digitalWrite(PIN_MSG_LED, LOW);

}



/*** Metodo para la lectura de los datos del raton PS/2 ***/
void MainCode::ReadMouse() {

    // Lectura de los datos actuales
    ps2mouse.GetReport(mouse.data);

    // Actualiza el desplazamiento del cursor
    mouse.position.x += mouse.data[1];
    mouse.position.y -= mouse.data[2];

    // Actualiza la rueda (si esta presente)
    mouse.position.wheel += mouse.data[3];

    // Boton Izquierdo
    if (mouse.data[0] & 0x01) {
        mouse.button.left |= true;
        mouse.event |= true;
    }

    // Boton Derecho
    if (mouse.data[0] & 0x02) {
        mouse.button.right |= true;
        mouse.event |= true;
    }

    // Boton Central
    if (mouse.data[0] & 0x04) {
        mouse.button.middle |= true;
        mouse.event |= true;
        // Compensacion de retardo por la lectura de paquetes extra al pulsar el boton central
        if (mouse.button.middle && !mouse.last.middle) ps2mouse.EmptyBuffer();
    }

}



/*** Matodo para el analisis y preparacion de los datos para su envio ***/
void MainCode::ParseData() {

    // Rueda del raton
    mouse.delta.wheel = mouse.position.wheel;

    // Ajusta la velocidad del cursor con la rueda (si esta disponible)
    if (mouse.delta.wheel != mouse.last.wheel) {
        if (mouse.delta.wheel > 0) {
            mouse.speed += MOUSE_SPEED_CHANGE;
            if (mouse.speed > MOUSE_MAX_SPEED) mouse.speed = MOUSE_MAX_SPEED;
        }
        if (mouse.delta.wheel < 0) {
            mouse.speed -= MOUSE_SPEED_CHANGE;
            if (mouse.speed < MOUSE_MIN_SPEED) mouse.speed = MOUSE_MIN_SPEED;
        }
    }

    // Ajuste de velocidad
    mouse.delta.x = (mouse.position.x * mouse.speed);
    mouse.delta.y = (mouse.position.y * mouse.speed);

    // Si se detecta desplazamiento, ajusta los valores delta, reinicia los contadores y registra el evento
    if (mouse.delta.x != 0) {
        mouse.position.x = 0;
        mouse.event |= true;
        if (mouse.delta.x < -255) mouse.delta.x = -255;
        if (mouse.delta.x > 255) mouse.delta.x = 255;
    }
    if (mouse.delta.y != 0) {
        mouse.position.y = 0;
        mouse.event |= true;
        if (mouse.delta.y < -255) mouse.delta.y = -255;
        if (mouse.delta.y > 255) mouse.delta.y = 255;
    }

    // Captura el movimiento de la rueda
    if (mouse.delta.wheel != 0) {
        mouse.position.wheel = 0;
        mouse.event |= true;
        if (mouse.delta.wheel < -7) mouse.delta.wheel = -7;
        if (mouse.delta.wheel > 7) mouse.delta.wheel = 7;
    }

    // Captura los eventos por el cambio de estado al pulsar/soltar cualquier boton del mouse
    mouse.event |= (mouse.button.left != mouse.last.left);
    mouse.event |= (mouse.button.right != mouse.last.right);
    mouse.event |= (mouse.button.middle != mouse.last.middle);

}



/*** Metodo el LED de actividad del raton ***/
void MainCode::ActivityLed() {

    // Led de actividad
    if (mouse.event) {
        if (mouse.restore_flag) {
            digitalWrite(PIN_MSG_LED, HIGH);
            mouse.led = MOUSE_BLINK_DELAY;   
        } else if (mouse.save_flag) {
            digitalWrite(PIN_MSG_LED, LOW);
            mouse.led = MOUSE_BLINK_DELAY;
        } else if (mouse.button.left || mouse.button.right || mouse.button.middle || (mouse.delta.wheel != 0)) {
            digitalWrite(PIN_MSG_LED, HIGH);
            mouse.led = MOUSE_BLINK_DELAY;
        } else {
            mouse.led --;
            if (mouse.led <= 0) {
                mouse.led = MOUSE_BLINK_DELAY;
                digitalWrite(PIN_MSG_LED, !digitalRead(PIN_MSG_LED));
            }
        }
    } else {
        mouse.led = MOUSE_BLINK_DELAY;
        digitalWrite(PIN_MSG_LED, LOW);
    }

}



/*** Metodo para codificacion del paquete de datos ***/
void MainCode::EncodeData() {

    // Si no hay eventos registrados sal
    if (!mouse.event) return;

    // Prepara los datos de codificacion
    uint8_t x = mouse.delta.x;
    uint8_t y = mouse.delta.y;
    uint8_t rb = mouse.button.right ? 1 : 0;
    uint8_t lb = mouse.button.left ? 1 : 0;
    uint8_t mb = mouse.button.middle ? 1 : 0;
    mouse.packet_size = 3;

    // Codificacion del byte 0 del paquete
    mouse.packet[0] =
        ((x >> 6) & 0x03)                   // Ultimos 2 bits del delta X
        |
        (((y >> 6) & 0x03) << 2)            // Ultimos 2 bits del delta Y
        |
        ((rb << 4) | (lb << 5) | 0x40)      // Estados de los botones derecho e izquierdo y el ID de paquete
    ;

    // Codificacion del byte 1 del paquete
    mouse.packet[1] = (x & 0x3F);           // Primeros 6 bits del delta X
    // Codificacion del byte 2 del paquete
    mouse.packet[2] = (y & 0x3F);           // Primeros 6 bits del delta Y

    // Codificacion del byte 3 del paquete (estado del boton central)
    if (mouse.button.middle) {      // Envia el 4º paquete mientras el boton central este presionado
        mouse.packet[3] = 0x20;
        mouse.packet_size = 4;
    } else {                        // Envia el 4º paquete una vez cuando se suelte el boton central
        if (mouse.last.middle && !mouse.button.middle) {
            mouse.packet[3] = 0x00;
            mouse.packet_size = 4;
        }
    }

}



/*** Metodo para el envio del paquete de datos al puerto serie ***/
void MainCode::SendData() {

    // Fase 1 - Envio del paquete de datos en caso de evento del raton
    if (mouse.event) Serial.write(mouse.packet, mouse.packet_size);

    // Fase 2 - Envio de un paquete adicional en caso de evento del boton central
    if (mouse.button.middle != mouse.last.middle) {
        mouse.packet[0] = 0x40;
        mouse.packet[1] = 0x00;
        mouse.packet[2] = 0x00;
        mouse.packet_size = 3;
        Serial.write(mouse.packet, mouse.packet_size);
    }

}



/*** Metodo para el monitoreo del estado de la señal RTS del puerto serie ***/
void MainCode::RtsStatusMonitor() {

    if (digitalRead(PIN_RTS_SIGNAL)) {

        // Si no se ha recibido la señal previamente
        if (!rs232.rts) {

            // Envia los bytes de inicializacion del protocolo
            delay(14);
            Serial.write(0x4D);         // char: 'M'
            delay(63);
            Serial.write(0x33);         // char: '3'

            // Reinicia el estado de los botones
            mouse.button.left = false;
            mouse.button.right = false;
            mouse.button.middle = false;

            // Reinicia el registro de posicion
            mouse.position.x = 0;
            mouse.position.y = 0;
            mouse.position.wheel = 0;

            // Registra el cambio de flag del RTS
            rs232.rts = true;

            // Indicalo en el led de activadad
            digitalWrite(PIN_MSG_LED, HIGH);
            delay(500);
            digitalWrite(PIN_MSG_LED, LOW);

        }

    } else {

        rs232.rts = false;      // Flag de señal recibida abajo

    }

}


/*** Guarda la configuracion de velocidad ***/
void MainCode::SaveMouseSpeed() {

    // Si no estan presionados ambos botones, reinicia los temporizadores y sal
    if (!mouse.button.left || !mouse.button.right) {
        mouse.save_timer = (millis() + SAVE_HELD_TIME);
        mouse.restore_timer = (millis() + RESTORE_HELD_TIME);
        // Reinicia flags de guardado si se sueltan ambos botones
        if (!mouse.button.left && !mouse.button.right) {
            mouse.save_flag = false;
            mouse.restore_flag = false;
        }
        // Sal
        return;
    }

    // Datos
    bool save_now = false;
    uint8_t data = 0;

    // Guardado de los parametros actuales
    if (!mouse.save_flag) {
        if ((millis() >= mouse.save_timer)) { 
            data = (uint8_t)(mouse.speed * 100.0f);
            mouse.save_flag = true;
            save_now = true;
        }
    }

    // Reinicio y guardado de los parametros
    if (!mouse.restore_flag) {
        if ((millis() >= mouse.restore_timer)) { 
            mouse.speed = MOUSE_DEFAULT_SPEED;
            data = (uint8_t)(mouse.speed * 100.0f);
            mouse.restore_flag = true;
        }
    }

    // Guarda los parametros en la EEPROM si se solicita
    if (save_now) {
        EEPROM.write(0x00, CHK0);
        EEPROM.write(0x01, data);
        EEPROM.write(0x02, CHK2);
    }

}



/*** Lee la configuracion de velocidad ***/
void MainCode::LoadMouseSpeed() {

    uint8_t b0 = EEPROM.read(0x00);
    uint8_t data = EEPROM.read(0x01);
    uint8_t b2 = EEPROM.read(0x02);

    if ((b0 != CHK0) || (b2 != CHK2)) {
        // Si los datos son invalidos, crealos por defecto
        data = (uint8_t)(MOUSE_DEFAULT_SPEED * 100.0f);
        EEPROM.write(0x00, CHK0);
        EEPROM.write(0x01, data);
        EEPROM.write(0x02, CHK2);
    }

    mouse.speed = ((float)data / 100.0f);
    if (mouse.speed < MOUSE_MIN_SPEED) mouse.speed = MOUSE_MIN_SPEED;
    if (mouse.speed > MOUSE_MAX_SPEED) mouse.speed = MOUSE_MAX_SPEED;

}
